### ISSUE TITLE
Collect paginated timemaps

### ISSUE THREAD

**Issue by ibnesayeed** [oduwsdl/memgator#33]
Some archives have pagination in place where all the pages needs to be collected to prepare full TimeMap.

**Comment by machawk1** [comment]
This seems like a pretty major bug. Do you have a list of which archives are currently supplying paginated TimeMaps? This might be something to include in the archives.json. I think LANL has this archive metadata in their more verbose archives listing.

**Comment by ibnesayeed** [comment]
Yes, LANL has paging information available in their [Archive registry](http://labs.mementoweb.org/aggregator_config/archivelist.xml). For IA it says, "timemaps are paged chronologically forward" which I see the response being sorted in chronologically forward order, but I don't see any paging there even when response has 50K+ Mementos. So, I am relaxed that it is not affecting the overall number of Mementos that we are aggregating.

As far as inclusion of the sort order and paging information in the archives list is concerned, it can be helpful in some optimization tricks, but otherwise I don't see much utility of it. This will increase the overhead of each archive entry signature and will become complex and confusing for users who will be adding custom archives list. Additionally, if an archive silently changes the behavior, detecting such changes and fixing them timely will be difficult.

The current implementation is optimized for "chronologically forward" order which is the default behavior of the OpenWayback and many big archives return their responses that way, but it can even handle unsorted responses (because sorting is not enforced by the RFC).

As far as paging is concerned, one approach to implement that is to look for `rel="timemap"` with `from` and `until` attributes while parsing a TimeMap and push the discovered URI into a session set to keep track of previously discovered pages. If the newly discovered page is not already requested, a new go routine can be instantiated to fetch that. This approach will even handle Index TimeMaps. To test the functionality, LANL's aggregators (DIY and WDI) can be used as source archives.

**Comment by galsondor** [comment]
I started dealing with pages timemaps about 18 months ago.  Both the Internet Archive and LANL aggregator conform to RFC 7089 § 5.1.1 Paging Timemaps (https://tools.ietf.org/html/rfc7089#section-5.1.1).  Page size for both is 1,000 mementos.

#### Recommendations:
* Memgator should figure out how implement paging or risk timing out trying to build one huge timemap.  For example on 2014-10-12, both ga.js and urchin.js (part of google analytics) had over 100,000 mementos each with a combined timemap size of ~1.2MB.  (I have encountered an image with over 200,000 mementos, but don't recall the URI-R right now).  The elapsed time to retrieve all 110 timemap pages was over 9 minutes.

#### Notes:
* Although both IA and LANL ostensibly order by Memento-Datetime, I did encounter instances where this order was violated.
* In the case where there are multiple mementos for the same Memento-Datetime, they can be split across multiple pages.

**Comment by ibnesayeed** [comment]
Thanks for the feedback @galsondor. Not having pagination simplifies a lot of things in usual client applications, that's why I wanted to return the complete response as the primary focus of the tool, especially, when it is encouraged to use it locally. Additionally, without local caching (in the form of marshaled object), pagination is a more costly business for the server. Pagination, if ever implemented would have a different approach in it which will be guided by the client. For example, the client may ask all the Mementos in a given time range. This approach will allow the MemGator server to identify when to stop asking for more pages from the upstream in some cases.

> Page size for both is 1,000 mementos.

I don't find it to be the case for IA now.

```
$ curl -s "http://web.archive.org/web/timemap/link/http://www.google.com/" | wc -l
50070
```

I also tried to grep for TimeMap pagination links in the response, but they are not present.

**Comment by galsondor** [comment]
@ibnesayeed Did a quick search of my research data cache.  Only the LANL aggregator returned paged timemaps during the Oct 2014 run.  Since then a new LANL aggregator has been implemented; it returns indexed timemaps vice paged timemaps.

Note that 'ga.js' is now 149 pages.

```sh
$ curl 'http://labs.mementoweb.org/timemap/link/http://www.google-analytics.com/ga.js' \
  | fgrep 'rel="timemap"' | wc -l
149
```
The first ```curl -I``` attempt returned a 503 Gateway timeout, probably due to the amount of data accumulated before the index timemap could be constructed.  So, it is easy to infer that collecting paged or index timemaps will create similar timing challenges for memgator.
